core/api/rpc/ws.cpp

```diff
@@ -136,9 +137,105 @@ namespace fc::api {
     Rpc rpc;
   };
 
+  struct HttpSession : public std::enable_shared_from_this<HttpSession> {
+    HttpSession(tcp::socket &&socket, std::shared_ptr<Api> api, Routes routes)
+        : socket(std::move(socket)),
+          timer(this->socket.get_executor(),
+                (std::chrono::steady_clock::time_point::max)()),
+          routes(std::move(routes)),
+          api{std::move(api)} {}
+
+    void run() {
+      onTimer({});
+
+      doRead();
+    }
+
+    void onTimer(boost::system::error_code ec) {
+      if (ec && ec != boost::asio::error::operation_aborted) return;
+
+      if (ec == boost::asio::error::operation_aborted) {
+        socket.shutdown(tcp::socket::shutdown_both, ec);
+        socket.close(ec);
+        return;
+      }
+
+      timer.async_wait([self{shared_from_this()}](beast::error_code ec) {
+        self->onTimer(ec);
+      });
+    }
+
+    void doRead() {
+      request = {};
+
+      http::async_read(
+          socket,
+          buffer,
+          request,
+          [self{shared_from_this()}](beast::error_code ec,
+                                     std::size_t bytes_transferred) {
+            self->onRead(ec);
+          });
+    }
+
+    void onRead(boost::system::error_code ec) {
+      if (ec == http::error::end_of_stream) return doClose();
+
+      if (ec) return;
+
+      if (websocket::is_upgrade(request)) {
+        std::make_shared<SocketSession>(std::move(socket), *api)
+            ->doAccept(std::move(request));
+        return;
+      }
+
+      handleRequest();
+    }
+
+    void handleRequest() {
+      for (auto &route : routes) {
+        if (request.target().starts_with(route.first)) {
+          route.second(request, response);
+          writeResponse();
+          break;
+        }
+      }
+    }
+
+    void writeResponse() {
+      response.content_length(response.body().size());
+
+      http::async_write(
+          socket,
+          response,
+          [self = shared_from_this()](beast::error_code ec, std::size_t) {
+            self->socket.shutdown(tcp::socket::shutdown_send, ec);
+          });
+    }
+
+    void doClose() {
+      boost::system::error_code ec;
+      socket.shutdown(tcp::socket::shutdown_send, ec);
+    }
+
+    // TODO: maybe add queue for requests
+
+    tcp::socket socket;
+    boost::beast::flat_buffer buffer;
```

Comment:
Why do we need boost:: here if you already defined beast namespace in .hpp file?

Reply:
Fixed